### DISCUSSION TITLE
Settings for ModuleDefinitions

### DISCUSSION THREAD

**Discussion by leigh-pointer** [oqtane/oqtane.framework#1829] [Ideas]
I think a nice extension to the ISettingService would be the ability to Add Update settings for the ModuleDefinition.  These settings are regardless to the module instance but more for the module itself.  i.e.. for a ExternMod i want to create some settings, these settings are used for all instances of the module.

`SettingService.GetModuleDefinitionSettingsAsync(ModuleState.ModuleDefinition.ModuleDefinitionId);`

`await SettingService.UpdateModuleDefinitionSettingsAsync(settings, ModuleState.ModuleDefinition.ModuleDefinitionId);
`

Thoughts?

**Comment by sbwalker** [comment]
@leigh-pointer the Settings area in Oqtane was architected in a generic manner so that it can be used as an attribute store for any entity - core or third party. The SettingService provides a convenient wrapper for the most common core entities, and I think adding support for ModuleDefinition makes sense. In addition to the methods outlined above you will need to add some logic to the SettingController in the IsAuthorized method to ensure that only users in the Host role are able to access these methods.

**Reply by leigh-pointer** [reply]
@sbwalker i had already started a local branch for this so i will see how i go later this evening.

**Comment by sbwalker** [comment]
It should require very minimal code changes to enable this if you follow the existing patterns.  And the shared enum EntityNames already contains ModuleDefinition.

**Reply by leigh-pointer** [reply]
@sbwalker testing now ;)

**Reply by Justincale** [reply]
Hi all, any news on this? I am creating a custom module which needs additional, custom settings. Thanks

**Reply by sbwalker** [reply]
@Justincale support for ModuleDefinition Settings was added ~2 years ago. You can access it via the SettingService.

**Reply by Justincale** [reply]
Sorry Shaun, I misread this discussion topic. What I require is to create and update module instance settings. Shall I create a new discussion?

**Reply by sbwalker** [reply]
Module (instance) Settings have been supported since the first version of Oqtane. If you created your module using the Create Module scaffolding feature in the framework then it already contains module settings as an example. Or you could look at the DNF.Projects module repo example (or most other examples). Settings are generally managed with a Settings.razor component and the SettingService API (which handles ALL types of settings in Oqtane)
